packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,290 @@
+            } else if (action == 5) { // OPEN_MAKE 
+                address product; UFixed18 amount;
+                (product, amount, ptr) = decodeAddressAmount(input, ptr);
+
+                IProduct(product).openMakeFor(msg.sender, amount);
+            } else if (action == 6) { // CLOSE_MAKE
+                address product; UFixed18 amount;
+                (product, amount, ptr) = decodeAddressAmount(input, ptr);
+
+                IProduct(product).closeMakeFor(msg.sender, amount);
+            } else if (action == 7) { // CLAIM 
+                address product; uint256[] memory programIds;
+                (product, programIds, ptr) = decodeProductPrograms(input, ptr);
+
+                controller.incentivizer().claimFor(msg.sender, IProduct(product), programIds);
+            } else if (action == 8) { // WRAP 
+                address receiver; UFixed18 amount;
+                (receiver, amount, ptr) = decodeAddressAmount(input, ptr);
+
+                wrap(receiver, amount);
+            } else if (action == 9) { // UNWRAP
+                address receiver; UFixed18 amount;
+                (receiver, amount, ptr) = decodeAddressAmount(input, ptr);
+
+                unwrap(receiver, amount);
+            } else if (action == 10) { // WRAP_AND_DEPOSIT
+                address account; address product; UFixed18 amount;
+                (account, product, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+                
+               wrapAndDeposit(account, IProduct(product), amount);
+            } else if (action == 11) { // WITHDRAW_AND_UNWRAP
+                address receiver; address product; UFixed18 amount;
+                (receiver, product, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                withdrawAndUnwrap(receiver, IProduct(product), amount);
+            } else if (action == 12) { // VAULT_DEPOSIT
+                address depositer; address vault; UFixed18 amount;
+                (depositer, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultDeposit(depositer, IPerennialVault(vault), amount);
+            } else if (action == 13) { // VAULT_REDEEM
+                address vault; UFixed18 shares;
+                (vault, shares, ptr) = decodeAddressAmount(input, ptr);
+
+                IPerennialVault(vault).redeem(shares, msg.sender);
+            } else if (action == 14) { // VAULT_CLAIM
+                address owner; address vault;
+                (owner, vault, ptr) = decodeAddressAddress(input, ptr);
+
+                IPerennialVault(vault).claim(owner);
+            } else if (action == 15) { // VAULT_WRAP_AND_DEPOSIT 
+                address account; address vault; UFixed18 amount;
+                (account, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultWrapAndDeposit(account, IPerennialVault(vault), amount);
+            }
+        }
+    }
+
+    /// Example Calldata Structure
```

Comment:
This can be moved to the `@dev` comment of the contract definition up at the top of the file

Reply:
sg needs updating too